File: backend/common/security/src/main/java/com/veda/server/JwtService.java
package com.veda.server;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import javax.crypto.SecretKey;
import java.util.Base64;
import java.util.Date;

@Service
public class JwtService {

    private final SecretKey secretKey;
    private final long accessTokenValidityMs = 15L * 60L * 1000L;
    private final long refreshTokenValidityMs = 30L * 24L * 60L * 60L * 1000L;

    public JwtService(@Value("${jwt.secret}") String jwtSecret) {
        byte[] decodedKey = Base64.getDecoder().decode(jwtSecret);
        this.secretKey = Keys.hmacShaKeyFor(decodedKey);
    }

    public String generateAccessToken(String userId) {
        return generateToken(userId, "access", accessTokenValidityMs);
    }

    public String generateRefreshToken(String userId) {
        return generateToken(userId, "refresh", refreshTokenValidityMs);
    }

    public boolean validateAccessToken(String token) {
        Claims claims = parseAllClaims(token);
        if (claims == null) return false;

        String tokenType = claims.get("type", String.class);
        return "access".equals(tokenType);
    }

    public boolean validateRefreshToken(String token) {
        Claims claims = parseAllClaims(token);
        if (claims == null) return false;

        String tokenType = claims.get("type", String.class);
        return "refresh".equals(tokenType);
    }

    public String getUserIdFromToken(String token) {
        Claims claims = parseAllClaims(token);
        if (claims == null) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid token");
        }
        return claims.getSubject();
    }

    public long getAccessTokenValidityMs() {
        return accessTokenValidityMs;
    }

    public long getRefreshTokenValidityMs() {
        return refreshTokenValidityMs;
    }

    private String generateToken(String userId, String type, long expiry) {
        Date now = new Date();
        Date expiryDate = new Date(now.getTime() + expiry);

        return Jwts.builder()
                .subject(userId)
                .claim("type", type)
                .issuedAt(now)
                .expiration(expiryDate)
                .signWith(secretKey, Jwts.SIG.HS256)
                .compact();
    }

    private Claims parseAllClaims(String token) {
        String rawToken = token;
        if (token != null && token.startsWith("Bearer ")) {
            rawToken = token.substring(7);
        }

        try {
            return Jwts.parser()
                    .verifyWith(secretKey)
                    .build()
                    .parseSignedClaims(rawToken)
                    .getPayload();
        } catch (Exception e) {
            return null;
        }
    }
}

File: src/main/kotlin/com/titsuko/server/security/JwtService.kt
package com.titsuko.server.security

import com.titsuko.server.dto.response.AuthResponse
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.Base64
import java.util.Date

@Service
class JwtService(
    @Value($$"${jwt.secret}") private val jwtSecret: String
) {
    private val secretKey = Keys.hmacShaKeyFor(Base64.getDecoder().decode(jwtSecret))
    private val accessExpiry = 15L * 60

    fun generateAccessToken(email: String): Pair<String, Long> {
        val now = System.currentTimeMillis()
        return Jwts.builder()
            .subject(email)
            .issuedAt(Date(now))
            .expiration(Date(now + accessExpiry * 1000))
            .signWith(secretKey)
            .compact() to accessExpiry
    }

    fun extractEmail(token: String): String? {
        return getClaims(token)?.subject
    }

    fun isTokenValid(token: String): Boolean {
        val claims = getClaims(token) ?: return false
        return !claims.expiration.before(Date())
    }

    private fun getClaims(token: String): Claims? {
        return try {
            Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .payload
        } catch (e: Exception) {
            null
        }
    }
}
